trunk/NDHMS/Routing/Reservoirs/module_read_timeslice_data.F

```diff
@@ -0,0 +1,354 @@
+
+            gage_id_string = gage_ids_string_array(timeslice_gage_index)
+
+            gage_id_string_trimmed = ADJUSTL(trim(gage_id_string))
+
+            number_counter = 0
+
+            do char_index = 1, 15
+                temp_char = gage_id_string_trimmed(char_index:char_index)
+
+                if (temp_char .eq. '0' .or. temp_char .eq. '1' .or. temp_char .eq. '2' &
+                .or. temp_char .eq. '3' .or. temp_char .eq. '4' .or. temp_char .eq. '5' &
+                .or. temp_char .eq. '6' .or. temp_char .eq. '7' .or. temp_char .eq. '8' &
+                .or. temp_char .eq. '9') then
+                    number_counter = number_counter + 1
+                end if
+
+            end do
+
+            read(gage_id_string_trimmed(1:number_counter), *) gage_id_integer
+
+            this%reservoir_gage_ids(timeslice_gage_index) = gage_id_integer
+
+        end do
+
+        if(allocated(gage_ids_string_array)) deallocate(gage_ids_string_array)
+
+        this%start_date = start_date
+        this%current_date = start_date
+        this%timeslice_path = timeslice_path
+        this%observation_lookback_hours = observation_lookback_hours
+
+        this%last_update_time_seconds = -999
+
+        ! Allocate lookback and dishcharge arrays
+        allocate (this%gage_lookback_seconds(this%number_of_reservoir_gages))
+        allocate (this%reservoir_gage_discharges(this%number_of_reservoir_gages))
+
+    end subroutine read_timeslice_data_struct_init
+
+
+    ! Timeslice Data Struct Destructor
+    subroutine read_timeslice_data_struct_destroy(this)
+
+        implicit none
+        class(read_timeslice_data_struct), intent(inout) :: this ! object being destroyed
+
+    end subroutine read_timeslice_data_struct_destroy
+
+
+    ! Set up list of timeslice files to read
+    ! The timeslices are read only once on each processor at each timeslice update time
+    subroutine setup_read_timeslice(this, update_interval_seconds, current_reservoir_time_seconds, &
+        reservoir_gage_id, gage_lookback_seconds, reservoir_gage_discharge)
+        implicit none
+
+        class(read_timeslice_data_struct), intent(inout) :: this
+        integer, intent(in)  :: update_interval_seconds
+        integer, intent(in)  :: current_reservoir_time_seconds
+        integer, intent(in)  :: reservoir_gage_id
+        integer, intent(out) :: gage_lookback_seconds
+        real, intent(out)    :: reservoir_gage_discharge
+        character(len=256)   :: timeslice_file_name
+        character(len=19)    :: old_date, new_date
+        character(len=2)     :: observation_resolution_string
+        integer :: total_timeslice_time_periods, timeslice_file_index, reservoir_subset_index, lake_index
+        integer :: observation_minute, observation_resolution_seconds
+        logical :: file_exists
+
+        ! This check ensures that the timeslice is read only once on each processor at each timeslice update time
```

Comment:
The file is being read by each processor? I thought it was being read once per node.

Reply:
Maybe i'm mis interpreting that

Reply:
From the documentation
> The first hybrid reservoir on each processor to reach an update time will call the function to read the timeslice files, and the resulting observations will be used by each reservoir on that processor. 

Does file contention become an issue at CONUS scale?

Reply:
I would have guessed that MPI would be faster than independent reads. But that's just a guess, i dont know that

Reply:
It is read by each processor. I updated the documentation to add more clarity:

"The timeslice files will be read at a particular update time. For a particular timestep, the first hybrid reservoir on each processor to reach an update time at that timestep will call the function to read the timeslice files, which will read the observations for every reservoir and store those values in an array. Each subsequent reservoir held by the same processor at that timestep that reaches its update time will read its corresponding observation from the array." 

Nels says that file contention shouldn't be an issue. We decided to go with this method because signaling one processor to read and broadcast the values and accounting for the various update times for each reservoir would be much more complex, though this process could be investigated in the next version.